### PR TITLE
Harden MSAAD sync (ASG-3350)

### PR DESCRIPTION
**Pull request by bosvos**
Fix deadlock bugs

### CODE REVIEW

File: msaad.go [hunk 1]
```diff
@@ -751,32 +756,58 @@ func numParallelFetchThreads(nItems int) int {
...
 		t := i % nThreads
 		threadGroups[t] = append(threadGroups[t], u)
 	}
 	wg := sync.WaitGroup{}
-	errChan := make(chan error, nThreads)
-
+	var errGlobal error
 	startTime := time.Now()
+	i := 0
 	for _, threadGroupOuter := range threadGroups {
```
Comment (FritzOnFire) [inline msaad.go:774]: Pretty sure you can just piggy back off this.
```suggestion
	for i, threadGroupOuter := range threadGroups {
```

File: msaad.go [hunk 2]
```diff
@@ -870,7 +899,9 @@ func (m *MSAAD) doHTTP(request *http.Request) (*http.Response, error) {
 
 	request.Header.Set("Authorization", "Bearer "+token)
 
-	return http.DefaultClient.Do(request)
+	client := http.DefaultClient
+	client.Timeout = 5 * time.Second
```
Comment (FritzOnFire) [inline msaad.go:903]: That is a tight deadline, I am going to assume there is a good reason for this.

File: auth.go [hunk 3]
```diff
@@ -733,7 +733,14 @@ func (x *Central) StartMergeTicker() error {
 	}
 	x.syncMergeTickerStopRequest = make(chan bool)
 	x.syncMergeTickerStopResponse = make(chan bool)
+
 	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				x.Log.Errorf("MergeTick panic: %v", r)
```
Comment (FritzOnFire) [inline auth.go:740]: So in mm/rad we have this whole `GetStack` function that we use when logging the error. Does the `r` have enough info that we don't need to dump the stack?

Here is the `GetStack` function for reference
```go
func GetStack() string {
	buf := make([]byte, 1024)
	for {
		n := runtime.Stack(buf, false)
		if n < len(buf) {
			return string(buf[:n])
		}
		buf = make([]byte, 2*len(buf))
	}
}
```

File: msaad.go [hunk 4]
```diff
@@ -913,6 +946,7 @@ func (m *MSAAD) getBearerToken() (token string, expiresAt time.Time, err error)
 	}
 
 	token = tokenJSON.AccessToken
+	// can time.Duration(string) not PANIC?
```
Comment (FritzOnFire) [inline msaad.go:949]: `tokenJSON.ExpiresIn` is an `int`. Go lang would not allow you to put anything in there that is not an `int`. Golang also sets default values.

**Review by FritzOnFire — APPROVED** [review]
Some questions and statements.

File: msaad.go (hunk 1 above)
Comment (bosvos) [reply msaad.go:774]: of course

File: msaad.go (hunk 2 above)
Comment (bosvos) [reply msaad.go:903]: Purely because there was no deadline in place prior...suggestions?

File: auth.go (hunk 3 above)
Comment (bosvos) [reply auth.go:740]: Good call...will implement and test.

Comment (bosvos) [reply auth.go:740]: Was not enough info initially (only high level error message, e.g. `division by zero`). Will implement GetStack as is...

**Review by FritzOnFire — APPROVED** [review]
LGTM

**Pull request merged**